feat(rfc): Add RFC Parsec auth (#9757)
Scille/parsec-cloud: docs/rfcs/1012-parsec-auth.md

# Save parsec device on a remote server

## Overview

This RFC will discuss the implementation of a new service that will keep the user's devices remotely while still ensuring that only the user can use them.

## Background & Motivation

During our reflexion about providing a client that could run in a web-browser came the questions:

- How should we store the device in the browser?.

  Some concern where raised about the persistance of a parsec device in the browser.

- What would the user experience be like?

  - What should happen when the user use a different browser?
  - What the user is expecting when using another machine?

  We assume that the user expect to be able to connect from any browser/machine.

That how we came to the idea of storing the device on a remote server:

- That solve the issue where the browser may clean it's data.
- The device would be accessible anywhere (given a network connection).

## Goals

How would a user would store/access its devices from a remote service.
What are the API that would be needed to implement this feature.

## Non-Goals

How to ensure that the client is not compromised.

## Design

> The authentification service would be a used before connecting to the parsec-cloud server to retrieve the device of the user that will then be used to authenticate to parsec-cloud.

Reviewer: ```suggestion
The authentification service would be used before connecting to the parsec-cloud server to retrieve the device of the user that will then be used to authenticate to parsec-cloud.
```

Author: done

2. Decrypt the key with `PRIV_AUTH_KEY`.
3. Encrypt the new device with `SYM_KEY` (`ENC_DEV`).
4. Upload the encrypted device (`ENC_DEV`) and the id of the organisation (`org_id`) to the server.

The server create a entry whose primary key is the tuple `user_id` and `org_id`.

> That mean a single device per user per organisation.

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

  Note over Alice,S: Alice is already authenticated to the server

  S ->> Alice: Retrieve the key `enc(SYM_KEY)` linked to `PUB_AUTH_KEY`
  Alice ->> Alice: Decrypt the key `enc(SYM_KEY)` with `PRIV_AUTH_KEY`
  Alice ->> Alice: Encrypt the device with `SYM_KEY` (`ENC_DEV`)
  Alice ->> S: Upload `ENC_DEV` + `org_id` to the server
```

### List available device on the service

Once authenticated, the user can list the device associated to the account.

> That list does not contain the blob `ENC_DEV`.

### Retrieve a device from the service

The server provide an API to retrieve the blob `ENC_DEV` of a device.
The user can then decrypt the blob on its side.

### Adding a new authentification method

> To add a new authentification method, it require the use to:

Reviewer: ```suggestion
To add a new authentification method, it require the user to:
```

Author: done

1. Create a new asymmetric key pair (`PUB_AUTH_KEY_B`, `PRIV_AUTH_KEY_B`) from the new method.
2. Retrieve the symmetric key `enc(SYM_KEY)` from the server.
3. Decrypt the key with `PRIV_AUTH_KEY`.
4. Encrypt the symmetric key with `PUB_AUTH_KEY_B` (`enc'(SYM_KEY)`).
5. Upload `enc'(SYM_KEY)`, `PUB_AUTH_KEY_B` and the secret of the new authentification method to the server.

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

  Note over Alice,S: Alice is already authenticated to the server

  Alice ->> Alice: Create a new asymmetric key pair (`PRIV_AUTH_KEY_B`/`PUB_AUTH_KEY_B`)
  S ->> Alice: Retrieve the key `enc(SYM_KEY)` related to `PUB_AUTH_KEY`
  Alice ->> Alice: Decrypt the key `enc(SYM_KEY)` with `PRIV_AUTH_KEY`
  Alice ->> Alice: Encrypt the key `SYM_KEY` with `PUB_AUTH_KEY_B` (`enc'(SYM_KEY)`)
  Alice ->> S: Upload `enc'(SYM_KEY)`, `PUB_AUTH_KEY_B` and the secret of the authentification method
```

### Removing an authentification method

> To remove an authentification method from a user, it require the user to:

Reviewer: ```suggestion
To remove an authentification method from a user, it requires the user to:
```

Author: done

    Alice ->> Alice: Encrypt the device with `SYM_KEY_B` (`ENC_DEV_B`)
  end
  loop For each `PUB_AUTH_KEY`
    Alice ->> Alice: Encrypt the key `SYM_KEY_B` with `PUB_AUTH_KEY` (`enc(SYM_KEY_B)`)
  end
  Alice ->> S: Upload the list of encrypted device `ENC_DEV_B`
  Alice ->> S: Upload the list of key `enc(SYM_KEY_B)`
  Alice ->> S: Remove the authentication method from the server
```

### Using the service alongside the parsec-cloud server

The parsec client would use the authentification service to store a special device, who will only be use to create a new local device.

That new device would be stored on the local storage and if the storage happen to be cleaned, the user could still re-create a new device from the special one stored on the service.

For already existing organisations and devices, the client could device to create and upload a special device to the service (cf [Uploading a new device](#uploading-a-new-device)).

## Alternatives Considered

N/A

## Operations

That require managing a new service that will store the devices of the user.

## Security/Privacy/Compliance

Consideration should be taken about the method used to save the device on the remote server.

> The service would need to identify the user so we will need to have some information about the user (email mostly).

Reviewer: Should we generate a parsec auth user id  ? That way a user may have the option to change their email.

Author: Indeed that would be better

Author: done

  For the password, it's not used as is, but it's derived to create a new secret using the function `PBKDF_A` (this operation is done on the client side to not leak the master password).
  The new secret will be provided as is to the server (like a password).

  In the Fido2 case we just use the included challenge since the private key is not shared with the server.

- An asymmetric key pair (`PUB_AUTH_KEY`, `PRIV_AUTH_KEY`)

  When using a password, the private key is generated by using the function `PBKDF_B` with the password

  > A key property here is we only need the password to generate the required keys.
  > Another approach would be to use the password as a symmetric key to encrypt a randomly generated private key that is sent alongside the public key.
  > But we would lose that property.

  For Fido2, the private already exists in the fido2 device so we use it directly.

  > [!NOTE]
  > Should verify if it's possible to use fido2 to encrypt random blob of data.

- A symmetric key (`SYM_KEY`) that is generated during the account creation at the end by the client.
  It's encrypted with `PUB_AUTH_KEY` and then uploaded to the service.

> [!CAUTION]

> > The output of `PBKDF_A` and `PBKDF_B` should be different to avoid leaking the private key.

Author: That done in the paragraph `### Consideration on PBKDF algorithms`

9. Remove the authentication method from the server.

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

  Note over Alice,S: Alice is already authenticated to the server

  S ->> Alice: Retrieve every key `PUB_AUTH_KEY` except the one to remove
  S ->> Alice: Get the symmetric key `enc(SYM_KEY)`
  S ->> Alice: Download every devices `ENC_DEV`
  Alice ->> Alice: Decrypt `enc(SYM_KEY)` with `PRIV_AUTH_KEY`
  Alice ->> Alice: Create a new symmetric key (`SYM_KEY_B`)
  loop For each device
    Alice ->> Alice: Decrypt the device with `SYM_KEY`
    Alice ->> Alice: Encrypt the device with `SYM_KEY_B` (`ENC_DEV_B`)
  end
  loop For each `PUB_AUTH_KEY`
    Alice ->> Alice: Encrypt the key `SYM_KEY_B` with `PUB_AUTH_KEY` (`enc(SYM_KEY_B)`)
  end
  Alice ->> S: Upload the list of encrypted device `ENC_DEV_B`
  Alice ->> S: Upload the list of key `enc(SYM_KEY_B)`
  Alice ->> S: Remove the authentication method from the server
```

### Using the service alongside the parsec-cloud server

The parsec client would use the authentication service to store a special device, who will only be used to create a new local device.

> That new device would be stored on the local storage and if the storage happens to be cleaned, the user could still re-create a new device from the special one stored on the service.

Reviewer 2: This section needs further details. We should explain a little bit why this "special device" and local devices are needed, why not directly using the special device to login, and _how_ exactly the special device is used to "create a new local device".

Also, we should probably give them more meaningful names. I think we've already called them "device registration key" (the one that is uploaded and allows to register new devices) and "device security key" (the "classic" Parsec device key).

Author: I've added more details why the device is not used as is

For already existing organizations and devices, the client could create and upload a special device to the service (cf [Uploading a new device](#uploading-a-new-device)).

## Alternatives Considered

N/A

## Operations

That require managing a new service that will store the devices of the user.

## Security/Privacy/Compliance

Consideration should be taken about the method used to save the device on the remote server.
The service would need to identify the user so we will need to have some information about the user (email mostly).

## Risks

> The devices are not stored securely to only allow the user to access them.

Author: I've added an explanation

  It's encrypted with `PUB_AUTH_KEY` and then uploaded to the service.

> [!CAUTION]
> The output of `PBKDF_A` and `PBKDF_B` should be different to avoid leaking the private key.

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

  Alice ->> S: Provide its email for the account creation
  S ->> Alice: Send an email with a unique code need to continue the process
  alt Choose between password or fido2
    Alice ->> Alice: Choose a password
    Alice ->> Alice: Derive a secret using the password and `PBKDF_A`
    Alice ->> Alice: Create `PRIV_AUTH_KEY`, `PUB_AUTH_KEY` using the password and `PBKDF_B`
  else
    Alice ->> Alice: Use the public key of the FIDO2 device as `PUB_AUTH_KEY`, the private key of FIDO2 device as `PRIV_AUTH_KEY`
  end

  Alice ->> Alice: Create the symmetric key `SYM_KEY`
  Alice ->> Alice: Encrypt the key `SYM_KEY` with `PUB_AUTH_KEY`
  Alice ->> S: Send the require data to finish the account creation (`enc(SYM_KEY)`, `PUB_AUTH_KEY`, `AUTH_SECRET`, email)
```

At the end of the process, the server should save the following information in a new entry:

> - A UUID (`user_id`)

Reviewer: Maybe make explicit that this user_id is not related to any organization's user_id.

Author: I've added a note to be more specific

- The email of the user
- The secret of the authentication method (`AUTH_SECRET`)

  > [!NOTE]
  > For password, it would be hashed using a strong hashing function.

- The public key of the authentication method (`PUB_AUTH_KEY`)
- The encrypted symmetric key (`enc(SYM_KEY)`).

> [!WARNING]
> A user should be able to register multiple authentication methods,
> so the user entry should allow for multiple `AUTH_SECRET`, `PUB_AUTH_KEY` and `enc(SYM_KEY)`.

### The authentication process

To be able to authenticate the user, it will need to provide the following information:

> - Its email

Reviewer: ```suggestion
- Their email
```

- The secret for the chosen authentication method (`AUTH_SECRET`)

The server then verifies the provided information and return a token `AUTH_TOKEN` if the information is correct.
That token would need to be provided to each request requiring to be authenticated.

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

  Alice ->> S: Provide its email and the authentication method secret
  S ->> S: Valide the provided information
  alt The information are correct
    S ->> Alice: Return an authentification token `AUTH_TOKEN`
  else The information are not correct
    S ->> Alice: Return a standard error for invalid authentification
  end
```

### Uploading a new device

To upload a new device, the user would need to:

0. Authenticate on the server to obtain the `AUTH_TOKEN` (see previous section).
1. Retrieve the symmetric key `enc(SYM_KEY)` from the server.
2. Decrypt the key with `PRIV_AUTH_KEY`.
3. Encrypt the new device with `SYM_KEY` (`ENC_DEV`).
4. Upload the encrypted device (`ENC_DEV`) and the ID of the organization (`org_id`) to the server.

> The server creates an entry whose primary key is the tuple `user_id` and `org_id`.

Reviewer: Just to be sure, it's the auth server user_id ? Not the organization user_id ?

Author: I've added a note to clarify that

- Provide `AUTH_SECRET` (so to be authenticated).
- Provide the secret of the authentication method.

  > This is to mimic "Enter your password to confirm the deletion of your account".

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

  Note over Alice,S: Alice is already authenticated to the server

  Alice ->> S: Provide the session token `AUTH_TOKEN` and the secret of the authentication method
  S ->> S: Remove the account and it's associated data (stored devices)
```

> [!IMPORTANT]
> We only delete the account in parsec-auth, no data are deleted in the metadata server.

### Integration with the Parsec server

The parsec client would use the authentication service to store a special device (let's call it remote device), who will only be used to create a new local device.
That new device would be stored on the local storage and if the storage happens to be cleaned, the user could still re-create a new device from the remote device.

We could directly use the remote device to communicate with the metadata server, but we choose the same approach used by the recovery device where it's only used to create new devices.

> One advantage is it provide a list of connected browsers/computers.

Author: > BTW the list of linked browsers

The ability to list the devices for a user (in a orga) is already implemented in the desktop

```mermaid
sequenceDiagram
  autonumber
  participant Client
  participant S_Auth as Authentification server
  participant S_Data as Metadata server

  Client  ->> S_Auth: Retrieve the device of the user
  S_Auth -->> Client: Device of the user

  Client  ->> S_Data: Exchange with the server
```

### Account creation

Creating the account for a user will require some information for the system to work:

- An identifier for the user (email)

  The service needs to verify that the user has access to the email (using a code or sending a link to continue the process).

- A HMAC shared key to authenticate the user (`AUTH_MEDIUM_HMAC_KEY`)

  When using a password, it's not used as is, but it's derived to create a HMAC key using the function `PBKDF_A` (this operation is done on the client side to not leak the master password).
  The new secret will be provided as is to the server (like a password).

  For the FIDO2 case the included challenge is used as is since the private key is not shared with the server.

- An asymmetric key pair (`AUTH_MEDIUM_PUB_KEY`, `AUTH_MEDIUM_PRIV_KEY`)

  They're generated using a cryptographic secure pseudo-random number generator (CSPRNG).

>   > This step is not needed when using a fido2 device since, it provides the asymmetric keys securely.

Reviewer: ```suggestion
  > This step is not needed when using a fido2 device since it provides the asymmetric keys securely.
```

  When using a password, it's generated from the function `PBKDF_B`

  For FIDO2, we can bypass that step since we do not share `AUTH_MEDIUM_PRIV_KEY` since it's stored on the device.

- A randomly generated symmetric key `ACCOUNT_MANIFEST_SYM_KEY`, it will be used to encrypt the account manifest.

- Generate the account manifest `ACCOUNT_MANIFEST`,

  the manifest will store:

  - The list of authentication public keys
  - The list of keys used to encrypt the devices (the list is empty at the beginning)

  > The manifest will be encrypted to prevent the server to temper with it.

> [!CAUTION]
> The output of `PBKDF_A` and `PBKDF_B` should be different to avoid leaking the private key.

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

  Alice ->> S: Provide its email for the account creation
  S ->> Alice: Send an email with a unique code need to continue the process
  alt Choose between password or fido2
    Alice ->> Alice: Choose a password
    Alice ->> Alice: Derive `AUTH_MEDIUM_HMAC_KEY` using the password and `PBKDF_A`
    Alice ->> Alice: Create `AUTH_MEDIUM_SYM_KEY` using the password and `PBKDF_B`

>     Alice ->> Alice: Generate `AUTH_MEDIUM_PRIV_KEY`, `AUTH_MEDIUM_PUB_KEY`

Reviewer: is there a real difference between create and generate ? 

Is it because the symetric key uses an input , and the asymetric pair doesn't ?

    Alice ->> Alice: Encrypt `AUTH_MEDIUM_PRIV_KEY` with `AUTH_MEDIUM_SYM_KEY`
  else
    Alice ->> Alice: Generate `AUTH_MEDIUM_HMAC_KEY` using the fido2 device.
    Alice ->> Alice: Use the public key of the FIDO2 device as `AUTH_MEDIUM_PUB_KEY`, the private key of FIDO2 device as `AUTH_MEDIUM_PRIV_KEY`
  end

  Alice ->> Alice: Create a bundle containing `AUTH_MEDIUM_PUB_KEY` (`ACCOUNT_MANIFEST`)
  Alice ->> Alice: Create a symmetric key `ACCOUNT_MANIFEST_SYM_KEY`
  Alice ->> Alice: Encrypt the bundle `ACCOUNT_MANIFEST` with `ACCOUNT_MANIFEST_SYM_KEY` (`enc(ACCOUNT_MANIFEST)`)
  Alice ->> Alice: Encrypt `ACCOUNT_MANIFEST_SYM_KEY` with `AUTH_MEDIUM_PUB_KEY` (`enc(ACCOUNT_MANIFEST_SYM_KEY)`)
  Alice ->> S: Send the require data to finish the account creation (`enc(AUTH_MEDIUM_PRIV_KEY)`, `AUTH_MEDIUM_SYM_KEY_PARAMS`, `enc(ACCOUNT_MANIFEST)`, `enc(ACCOUNT_MANIFEST_SYM_KEY)`, `AUTH_MEDIUM_HMAC_KEY`, email validation token)
```

At the end of the process, the server should save the following information:

- An entry for the user account:

>   - A UUID (`user_id`)

Reviewer: nit: maybe it could be renamed to account_id to avoid confusion.

Author: done

    > It's not related to a `user_id` inside a parsec organization.

  - The email of the user

  - The encrypted account manifest (`enc(ACCOUNT_MANIFEST)`)

- An entry for the authentication method linked to the user:

  - The HMAC secret (`AUTH_MEDIUM_HMAC_KEY`)
  - The encrypted private key (`enc(AUTH_MEDIUM_PRIV_KEY)`)

    > Could be empty when the user uses a FIDO2 device.

  - The encrypted key used to decrypt the account manifest (`enc(ACCOUNT_MANIFEST_SYM_KEY)`)
  - The parameters used to generate the symmetric key (`AUTH_MEDIUM_SYM_KEY`) that encrypted `enc(AUTH_MEDIUM_PRIV_KEY)` (`AUTH_MEDIUM_SYM_KEY_PARAMS`)

> [!WARNING]
> A user should be able to register multiple authentication methods,

### The authentication process

We use a HMAC to authenticate each request to the server with the secret `AUTH_MEDIUM_HMAC_KEY` that was previously shared.

### Obtain the account manifest from the server

The server store the encrypted account manifest that contain the important information about the user.

> For the user to obtain the account manifest from the server it will do:

Reviewer: ```suggestion
For the user to obtain the account manifest from the server they will do the following:
```

5. Generate a device ownership proof (`DEVICE_OWNERSHIP_PROOF`):

   The proof is generated by using the signature key of the device to sign the blob `device_id` + `auth_medium_id` currently used to authenticate the request.

6. Upload the organization, the encrypted device, the device ownership proof, and the updated encrypted account manifest.

   > The server will ask the metadata server to verify the device ownership proof since it has access to the verify key of the device.

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

  Note over Alice,S: Alice is already authenticated to the server

  S ->> Alice: Obtain the account manifest from the server following the procedure
  Alice ->> Alice: Generate `DEVICE_SYM_KEY`
  Alice ->> Alice: Encrypt the device `DEVICE` with `DEVICE_SYM_KEY` (`enc(DEVICE)`)
  Alice ->> Alice: Add `DEVICE_SYM_KEY` to the account manifest
  Alice ->> Alice: Generate `DEVICE_OWNERSHIP_PROOF`
  Alice ->> S: Upload `org_id`, `enc(DEVICE)`, `DEVICE_OWNERSHIP_PROOF`, `enc(ACCOUNT_MANIFEST)`
```

> [!NOTE]
> The user can upload multiple devices for the same organization.

### List available devices

> Once authenticated, the user can list the devices associated to its account.

Reviewer: ```suggestion
Once authenticated, the user can list the devices associated to their account.
```

> The list contain the information about the device like:

Reviewer: ```suggestion
The list contains the information about the device like:
```

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

  Note over Alice,S: Alice is already authenticated to the server

  Alice ->> S: Remove the account from the server
```

> [!IMPORTANT]
> We only delete the account in parsec-auth, no data are deleted in the metadata server.

### Integration with the Parsec server

The parsec client would use the authentication service to store a special device (let's call it remote device), which will only be used to create a new local device.
The new local device would be stored on the local storage. If the storage happens to be cleaned, a new device could still be created from the remote device.

We could have directly used the remote device to communicate with the metadata server,
but this approach (similar to the one used to create devices from a recovery device)
has the advantage of allowing to provide a list of linked browsers/computers.

For existing devices (created before/without the authentication service) the client would need to create and upload a remote device to the service (cf [Uploading a new device](#uploading-a-new-device)) in order to access Parsec from a web environment.

> ### Global flow diagram

Reviewer: :heart:

That is how we came to the idea of storing the device on a remote server.

- That will solve the issue of the browser cleaning its data.
- The devices would be accessible from anywhere (given the user is online).

## Goals

1. Describe how to store/access the user devices on a remote service.
2. Propose an API to implement this feature.

## Non-Goals

- Discuss security concerns regarding the web client application.

## Design

### General principles

An authentication service (Parsec Auth) would be used to store user devices. This service will be requested before connecting to the metadata server (Parsec server) in order to retrieve a specific user device. The device will then be used to authenticate to the metadata server.

<!-- Device stored on a third-party service -->
```mermaid
sequenceDiagram
  autonumber
  participant Client
  participant S_Auth as Authentification server
  participant S_Data as Metadata server

  Client  ->> S_Auth: Retrieve the device of the user
  S_Auth -->> Client: Device of the user

  Client  ->> S_Data: Exchange with the server
```

### Account creation

Creating the account for a user will require some information for the system to work:

> - An identifier for the user (email)

Author: done

  The service needs to verify that the user has access to the email (using a code or sending a link to continue the process).

- A HMAC shared key to authenticate the user (`AUTH_MEDIUM_HMAC_KEY`)

  When using a password, it's not used as is, but it's derived to create a HMAC key using the function `PBKDF_A` (this operation is done on the client side to not leak the master password).
  The new secret will be provided as is to the server (like a password).

  For the FIDO2 case the included challenge is used as is since the private key is not shared with the server.

> - An asymmetric key pair (`AUTH_MEDIUM_PUB_KEY`, `AUTH_MEDIUM_PRIV_KEY`)

Author: done

  They're generated using a cryptographic secure pseudo-random number generator (CSPRNG).

  > This step is not needed when using a fido2 device since it provides the asymmetric keys securely.

- A symmetric key used to encrypt `AUTH_MEDIUM_PRIV_KEY` (`AUTH_MEDIUM_SYM_KEY`) alongside its parameters used to generate it `AUTH_MEDIUM_SYM_KEY_PARAMS`.

  When using a password, it's generated from the function `PBKDF_B`

>   For FIDO2, we can bypass that step since we do not share `AUTH_MEDIUM_PRIV_KEY` since it's stored on the device.

Reviewer 3: Regarding FIDO2, I think we are currently lacking actual experience to be able to correctly specify what is needed

For instance we know that we are going to need the FIDO2 extension that allows to generate a symmetric key.
It would be interesting to use this symmetric key as a password for our system, this way the FIDO2 implementation would only be a small layer on top of the password one.

So I think we should either
- focus on building a small POC showcasing how to use FIDO2 to create a symmetric key, doing auth, reseting auth etc.
- or dropping FIDO2 entirely from this RFC, and instead just consider we should support multiple authentication methods all based on a user-provided secret

Author: I understand that we lake experience concerning the usage of FIDO2 devices for authentication and such.

If I need to choose, it would prefer to drop reference to FIDO2 from the RFC for it to be describe in a future RFC

Reviewer 2: I agree we lack actual experience but at the same time FIDO2 was already discussed (analyzed?) during the study to support ProConnect for Resana.

So I'm fine with keeping FIDO2 in stand-by for the moment and just consider multiple auth methods based on a user-provided secret. Just make sure that we are not closing the door to support FIDO2.

Author: For confirmation: We do not remove the FIDO2 references from this RFC ? @mmmarcos

Reviewer 2: @FirelightFlagboy I am not sure. I think it will be a pity to lose the information on what's already been analyzed. I guess it depends on how relevant the information is and how much do you think it would be useful in the future.

So choose one of the following: 

1. Keep references to FIDO2 --> Add a note in the Overview section (or non goals section) based on @touilleMan comment:
```
Mentions to FIDO2 as an alternative authentication method should not be 
taken as decision made in stone. We are currently lacking actual experience 
to be able to  correctly specify what is needed. For instance a FIDO2 
extension allows to generate a symmetric key that could be used as password 
(so FIDO2 implementation would only be a small layer on top of the password).
```

2. Remove references to FIDO2 to have a "cleaner" RFC --> Move them somewhere (issue? draft rfc?).

Author: I've added the note in the overview section

- A randomly generated symmetric key `ACCOUNT_MANIFEST_SYM_KEY`, it will be used to encrypt the account manifest.

- Generate the account manifest `ACCOUNT_MANIFEST`,

  the manifest will store:

  - The list of authentication public keys

>   - The list of keys used to encrypt the devices (the list is empty at the beginning)

Author: done

  > The manifest will be encrypted to prevent the server to temper with it.

> [!CAUTION]
> The output of `PBKDF_A` and `PBKDF_B` should be different to avoid leaking the private key.

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

>   Alice ->> S: Provide its email for the account creation

Author: Could be send at the last step like the other values

>   Alice ->> S: Provide its email for the account creation

Author: included at the last step

    Alice ->> Alice: Generate `AUTH_MEDIUM_PRIV_KEY`, `AUTH_MEDIUM_PUB_KEY`
    Alice ->> Alice: Encrypt `AUTH_MEDIUM_PRIV_KEY` with `AUTH_MEDIUM_SYM_KEY`
  else
    Alice ->> Alice: Generate `AUTH_MEDIUM_HMAC_KEY` using the fido2 device.
    Alice ->> Alice: Use the public key of the FIDO2 device as `AUTH_MEDIUM_PUB_KEY`
  end

  Alice ->> Alice: Create a bundle containing `AUTH_MEDIUM_PUB_KEY` (`ACCOUNT_MANIFEST`)
  Alice ->> Alice: Create a symmetric key `ACCOUNT_MANIFEST_SYM_KEY`
  Alice ->> Alice: Encrypt the bundle `ACCOUNT_MANIFEST` with `ACCOUNT_MANIFEST_SYM_KEY` (`enc(ACCOUNT_MANIFEST)`)
  Alice ->> Alice: Encrypt `ACCOUNT_MANIFEST_SYM_KEY` with `AUTH_MEDIUM_PUB_KEY` (`enc(ACCOUNT_MANIFEST_SYM_KEY)`)
  Alice ->> S: Send the require data to finish the account creation<br>(`enc(AUTH_MEDIUM_PRIV_KEY)`, `AUTH_MEDIUM_SYM_KEY_PARAMS`, `enc(ACCOUNT_MANIFEST)`,<br>`enc(ACCOUNT_MANIFEST_SYM_KEY)`, `AUTH_MEDIUM_HMAC_KEY`, email validation token)
```

At the end of the process, the server should save the following information:

- An entry for the user account:

  - A UUID (`account_id`)

  - The email of the user

  - The encrypted account manifest (`enc(ACCOUNT_MANIFEST)`)

> - An entry for the authentication method linked to the user:

Reviewer 3: > Does it makes sense to add an authentication method (e.g. "password") multiple times? If not, how will the server prevent it?

I think we still want this (typically FIDO2 relies on the fact you end up with a different authentication source on each of the machine you want to connect from)

> If there are multiple authentication methods for an account, and the users wants to authenticate with "password", how will the server know which of the stored authentication methods to use?

The auth medium should have an ID derived from its secret that is send to the server as part of the request Authentication header ([similar to what is done for the authenticated API](https://github.com/Scille/parsec-cloud/blob/7e136748b4051ae47aeb092758e3feb83818cb1b/libparsec/crates/client_connection/src/authenticated_cmds/mod.rs#L357))

>   - A UUID (`auth_medium_id`)

Reviewer 3: I think the auth medium id should be derived from the secret, otherwise there is no way for the client to retrieve it !

Author: I've changed the UUID by an ID provided by the client with a note

  - The HMAC secret (`AUTH_MEDIUM_HMAC_KEY`)

>     The secret should be securely stored by either encrypting its value or using a vault.

Reviewer 3: This is not needed since this secret has a similar role to what would have a regular authentication token.

Author: The secret need to be store securely.
If we store it in the database and we have a leak someone could impersonate the user

  - The encrypted private key (`enc(AUTH_MEDIUM_PRIV_KEY)`)

    > Could be empty when the user uses a FIDO2 device.

  - The encrypted key used to decrypt the account manifest (`enc(ACCOUNT_MANIFEST_SYM_KEY)`)
  - The parameters used to generate the symmetric key (`AUTH_MEDIUM_SYM_KEY`) that encrypted `enc(AUTH_MEDIUM_PRIV_KEY)` (`AUTH_MEDIUM_SYM_KEY_PARAMS`)

> [!WARNING]
> A user should be able to register multiple authentication methods,

### The authentication process

We use a HMAC to authenticate each request to the server with the secret `AUTH_MEDIUM_HMAC_KEY` that was previously shared.

> ### Obtain the account manifest from the server

Reviewer 3: Regarding the account manifest, I think we need two things:

1. The account manifest should be versioned, as this would prevent concurrent overwritting
2. We should be able to create multiple independent account manifests
     - any given auth medium would be related to a single account manifest ID
     - any uploaded device could be related to multiple account manifest ID

Regarding 2, this would allow a clean way to recover once the password has been lost:
- the Parsec auth admin generate a password reset link and send it the user
- password reset link is used to create a brand new account manifest and a auth medium using a new password
- once the user is connected, he sees he currently has access to no device (since his account manifest is empty), but that he used to have access to device X/Y/Z that are protected by some old auth medium (typically showing something like `john@example.com on Coolorg, protected by a password configured on 2000-01-01T00:00:00Z`).
- The user can then try to use an old auth medium to get access the the previous auth medium, if he succeed then the content of the old account manifest is merged with the new one.
- Since we store by which account manifest a device is accessible by, we can hide from the recovery screen the devices present in older account manifests once they have been recovered.

The server store the encrypted account manifest that contain the important information about the user.
For the user to obtain the account manifest from the server it will do:

1. Be authenticated with the server.

> 2. Retrieve the following encrypted data from the sever:

Reviewer 3: Also note that since we authenticate the request with a given auth medium id, we should only be able to get back the data related to this specific auth medium.

Author: I explicitly indicated that the encrypted blob correspond to the current authentication method

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as  Authentication server

  Note over Alice,S: Alice is already authenticated to the server

  S ->> Alice: Retrieve `enc(ACCOUNT_MANIFEST)`, `enc(ACCOUNT_MANIFEST_SYM_KEY)`,<br>`enc(AUTH_MEDIUM_PRIV_KEY)` & `AUTH_MEDIUM_SYM_KEY_PARAMS`
  Alice ->> Alice: Generate `AUTH_MEDIUM_SYM_KEY` using the password and `AUTH_MEDIUM_SYM_KEY_PARAMS`
  Alice ->> Alice: Decrypt `enc(AUTH_MEDIUM_PRIV_KEY)` with `AUTH_MEDIUM_SYM_KEY`
  Alice ->> Alice: Decrypt `enc(ACCOUNT_MANIFEST_SYM_KEY)` with `AUTH_MEDIUM_PRIV_KEY`
  Alice ->> Alice: Decrypt `enc(ACCOUNT_MANIFEST)` with `ACCOUNT_MANIFEST_SYM_KEY`
```

### Uploading a new device

To upload a new device, the user would need to:

1. [Obtain the account manifest from the server](#obtain-the-account-manifest-from-the-server).
2. Create a new symmetric key for encrypting the device (`DEVICE_SYM_KEY`).
3. Encrypt the new device with `DEVICE_SYM_KEY` (`enc(DEVICE)`).
4. Add `DEVICE_SYM_KEY` to the account manifest.
5. Encrypt the account manifest with the symmetric key `ACCOUNT_MANIFEST_SYM_KEY`.
5. Generate a device ownership proof (`DEVICE_OWNERSHIP_PROOF`):

>    The proof is generated by using the signature key of the device to sign the blob `device_id` + `auth_medium_id` currently used to authenticate the request.

Reviewer 3: ```suggestion
   The proof is generated by using the signature key of the device to sign the blob `device_id` + `account_id` currently used to authenticate the request.
```

It is more logical to use account id since any other auth medium id sharing the same account id will be able to have access to this device

Author: The proof is only use to validate that the use can use the uploaded device that server not other purpose.
As such it's not save on the server side

> 6. Upload the organization, the encrypted device, the device ownership proof, and the updated encrypted account manifest.

Reviewer 3: what do you mean by "Upload the organization" ?

Author: I've added precision to the organization id

>    > The server will ask the metadata server to verify the device ownership proof since it has access to the verify key of the device.

Author: done

```

### Removing the account

If the user wants to remove its account, it requires to just send a request to the server.

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

  Note over Alice,S: Alice is already authenticated to the server

  Alice ->> S: Remove the account from the server
```

> [!IMPORTANT]
> We only delete the account in parsec-auth, no data are deleted in the metadata server.

### Integration with the Parsec server

The parsec client would use the authentication service to store a special device (let's call it remote device), which will only be used to create a new local device.
The new local device would be stored on the local storage. If the storage happens to be cleaned, a new device could still be created from the remote device.

We could have directly used the remote device to communicate with the metadata server,
but this approach (similar to the one used to create devices from a recovery device)
has the advantage of allowing to provide a list of linked browsers/computers.

> For existing devices (created before/without the authentication service) the client would need to create and upload a remote device to the service (cf [Uploading a new device](#uploading-a-new-device)) in order to access Parsec from a web environment.

Reviewer 3: > So the user never uploads the existing local devices (created before auth service)? It's always a new remote device that needs to be created and uploaded?

Yes we want to have dedicated device stored in parsec auth (and maybe later even mark them in their device certificate as only allowed to create other devices).

> BTW, the client will need to create and upload 1 remote device per existing local device? Or is it 1 remote device per organization?

Good point ! I think the client should simply check if the account manifest already contains a valid device.

Since we said that the device in parsec auth is a specific one that is only used for creating new device, then the process of connecting an existing device to a parsec auth account result in the same outcome no matter if the parsec auth account already contained or not a device: in both case we end up with a device in parsec auth that has the same organization ID / user ID than the device on the client, but has a different device ID.

> Currently, it was proposed that the creation (and upload) of a remote device was transparent for the user and happen on login. Should it be instead a manual operation triggered by the user?

I'm not sure what you mean by transparent... when the client does a login on an existing device store on the local machine, there is no parsec auth involved so nothing possible.

On the other hand, if the end-user choose to do parsec auth login on his client,  he will end up on a page listing:
- the devices stored on parsec auth and that no device on the machine has this user ID ==> a "connect" button is displayed for each one of them (and clicking it means we create a new device and store it on the machine)
- the devices stored on parsec auth and some device on the machine has this user ID ==> they are hidden
- the devices stored on the machine and whose user ID is on parsec auth ==> a "connect" button is displayed
- the devices stored on the machine and whose user ID is not on parsec auth (e.g. device created before parsec auth existed) ==> a "store in parsec auth" button[^1] is displayed (and clicking it upload a new device in parsec auth)

[^1]: there is no "connect" button since the device is not parsec of parsec auth, so if the user doesn't want to add it to parsec auth he should connect to it at the step he connected to parsec auth

> For existing devices (created before/without the authentication service) the client would need to create and upload a remote device to the service (cf [Uploading a new device](#uploading-a-new-device)) in order to access Parsec from a web environment.

Reviewer 2: What you call the "connect" button is what we were thinking to happen automatically when the user clicks on the organization (as if a login to org was done). So there are no distinction if the device is on Parsec auth or not: when a user clicks on the organization, Parsec will either "login to org" OR it will "connect + login to org".

This of course could be re-discussed, as front-end development is just about starting.

> For existing devices (created before/without the authentication service) the client would need to create and upload a remote device to the service (cf [Uploading a new device](#uploading-a-new-device)) in order to access Parsec from a web environment.

Author: Following an internal discussion with the front-end team, the behavior is not decided yet but the provided API will not corner us one way or another.

Since there is no actionable action from this discussion I will close it

    enc_PRIV_KEY
    enc_MANIFEST
    enc_MANIFEST_KEY
    enc_DEVICE
  end

  E[Email] & P[Password] & AUTH_MEDIUM_SYM_KEY_PARAMS --> H3{PBKDF_2} --> AUTH_SYM_KEY[AUTH_MEDIUM_SYM_KEY]

  AUTH_SYM_KEY & enc_PRIV_KEY --> D1{"Decrypt the private key"} --> PRIV_KEY[AUTH_MEDIUM_PRIV_KEY]

  PRIV_KEY & enc_MANIFEST_KEY --> D2{"Decrypt the account manifest symmetric key"} --> MANIFEST_KEY[ACCOUNT_MANIFEST_SYM_KEY]

  MANIFEST_KEY & enc_MANIFEST --> D3{"Decrypt the account manifest"} --> MANIFEST[ACCOUNT_MANIFEST]

  MANIFEST --> DEV_KEY[DEVICE_SYM_KEY]

  DEV_KEY & enc_DEVICE --> DEV[DEVICE]
```

## Alternatives Considered

N/A

## Operations

That require managing a new service that will store the devices of the user.

## Security/Privacy/Compliance

Consideration should be taken about the method used to save the device on the remote server.
The service would need to identify the user so we will need to have some information about the user (email mostly).

### Consideration on PBKDF algorithms

The `PBKDF_A` (used to derive a secret from the password) and `PBKDF_B` (used to derive the private key from the password) output should be different in order to avoid leaking the private key.

> To make the output different, we could:

Reviewer 3: Keep things simple: we already use argon2, so that's what we are going to use, with the salt composed of the email + a constant unique for each derivation.

Author: The goal was to note the different possibilities.
I've added our choice for using the salt method at the end of the list

- Use different algorithms (e.g.: `argon2id`, `scrypt`)
- Choose different salt (but they need to be deterministic from the information provided by the user only).
- Use different parameters (but that may reduce the security)

## Risks

The devices are not stored securely to only allow the user to access them.
In the case of using a password to login and to generate the asymmetric key pair,
consideration should be taken to ensure that we obtain the key pair from the derivation of the password.

In [Account creation](<#Account creation>), we use the password to generate 2 things:

- An intermediate secret using `PBKDF_A` provided as is to the server.
- The key pair using `PBKDF_B`.

We have the section [Consideration of PBKDF algorithms](<#Consideration on PBKDF algorithms>) that state some of the requirement.

## Remarks & open questions

> - The client application will need to communicate with the authentication service: should this be integrated into `libparsec`? Or it's the JS side that handle that?

Reviewer 3: Since we already do client/server communication in `libparsec`, I think it's straightforward to continue doing that ^^

Author: I've noted you reply in the RFC

- What to do when a user is revoked from an organization?

>   Should we remove the device, or should we just provide the information that the device for the given organization is revoked?

Reviewer 3: I think it is better to just inform  the user is revoked

This way, the question of removing a device stays orthogonal: even if the user is not revoked, we may want to remove the device from parsec auth (for instance if we have by mistake uploaded multiple devices for the same user ?)

Author: I've noted your reply in the RFC

    Alice ->> Alice: Derive `AUTH_METHOD_HMAC_KEY` using the password and `PBKDF_B`
    Alice ->> Alice: Create `AUTH_METHOD_SYM_KEY` using the password and `PBKDF_C`
    Alice ->> Alice: Generate `AUTH_METHOD_PRIV_KEY`, `AUTH_METHOD_PUB_KEY`
    Alice ->> Alice: Encrypt `AUTH_METHOD_PRIV_KEY` with `AUTH_METHOD_SYM_KEY`
  else
    Alice ->> Alice: Generate `auth_method_id` using fido2 device.
    Alice ->> Alice: Generate `AUTH_METHOD_HMAC_KEY` using the fido2 device.
    Alice ->> Alice: Use the public key of the FIDO2 device as `AUTH_METHOD_PUB_KEY`
  end

  Alice ->> Alice: Create a bundle containing `AUTH_METHOD_PUB_KEY` (`ACCOUNT_MANIFEST`)
  Alice ->> Alice: Create a symmetric key `ACCOUNT_MANIFEST_SYM_KEY`
  Alice ->> Alice: Encrypt the bundle `ACCOUNT_MANIFEST` with `ACCOUNT_MANIFEST_SYM_KEY` (`enc(ACCOUNT_MANIFEST)`)
  Alice ->> Alice: Encrypt `ACCOUNT_MANIFEST_SYM_KEY` with `AUTH_METHOD_PUB_KEY` (`enc(ACCOUNT_MANIFEST_SYM_KEY)`)
  Alice ->> S: Send the require data to finish the account creation
```

At the end of the process, the server should save the following information:

- An entry for the user account:

  - A UUID (`account_id`)

>   - User firstname
>   - User lastname

Author: **todo:** 
```suggestion
  - User username
```

  - The email of the user

- An entry for the account manifest

  - Identified with the provided `account_id`
  - A UUID (`account_manifest_id`)
  - Denoted as the first version of the manifest
  - The encrypted account manifest (`enc(ACCOUNT_MANIFEST)`)

- An entry for the authentication method linked to the user:

  - An ID that is provided by the user (`auth_method_id`)

    > That require the ID to be generated by the user with information available at the moment (likely derived from its secret).

  - The ID of the related account manifest (`account_manifest_id`)

  - The HMAC secret (`AUTH_METHOD_HMAC_KEY`)

    The secret should be securely stored by either encrypting its value or using a vault.

  - The encrypted private key (`enc(AUTH_METHOD_PRIV_KEY)`)

    > Could be empty when the user uses a FIDO2 device.

  - The encrypted key used to decrypt the account manifest (`enc(ACCOUNT_MANIFEST_SYM_KEY)`)
  - The parameters used to generate the symmetric key (`AUTH_METHOD_SYM_KEY`) that encrypted `enc(AUTH_METHOD_PRIV_KEY)` (`AUTH_METHOD_SYM_KEY_PARAMS`)

> [!WARNING]

> > A user should be able to register multiple authentication methods,

Author: **todo:** 
```suggestion
> A user should be able to register multiple authentication methods
```

6. Upload the encrypted & updated account manifest and the list of `enc(AUTH_METHOD_SYM_KEY')`.

   > The server will store the updated manifest as a new version

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

  Note over Alice,S: Alice is already authenticated to the server

  S ->> Alice: Get the account manifest from the server
  Alice ->> Alice: Remove the public key of the authentication method from the account manifest
  Alice ->> Alice: Create a new `ACCOUNT_MANIFEST_SYM_KEY'`
  loop For each public key in the manifest
    Alice ->> Alice: Encrypt `ACCOUNT_MANIFEST_SYM_KEY'` with the public key (`enc(ACCOUNT_MANIFEST_SYM_KEY')`)
  end
  Alice ->> S: Upload the encrypted & updated account manifest, the list of `enc(ACCOUNT_MANIFEST_SYM_KEY')`<br>and the ID of the authentication method to remove
```

### Removing the account

If the user wants to remove its account, it requires to just send a request to the server.

```mermaid
sequenceDiagram
  autonumber
  participant Alice
  participant S as Authentification server

  Note over Alice,S: Alice is already authenticated to the server

>   Alice ->> S: Remove the account from the server

Author: **todo:** 
```suggestion
  Alice ->> S: Remove the account from the server
  S ->> Alice: Send confirmation email with a URL to use to confirm the suppression
  
  Alice ->> S: Use the confirmation link to delete the account
```

    enc_DEVICE
  end

  E[Email] & P[Password] & AUTH_METHOD_SYM_KEY_PARAMS --> H3{PBKDF_2} --> AUTH_SYM_KEY[AUTH_METHOD_SYM_KEY]

  AUTH_SYM_KEY & enc_PRIV_KEY --> D1{"Decrypt the private key"} --> PRIV_KEY[AUTH_METHOD_PRIV_KEY]

  PRIV_KEY & enc_MANIFEST_KEY --> D2{"Decrypt the account manifest symmetric key"} --> MANIFEST_KEY[ACCOUNT_MANIFEST_SYM_KEY]

  MANIFEST_KEY & enc_MANIFEST --> D3{"Decrypt the account manifest"} --> MANIFEST[ACCOUNT_MANIFEST]

  MANIFEST --> DEV_KEY[DEVICE_SYM_KEY]

  DEV_KEY & enc_DEVICE --> DEV[DEVICE]
```

## Alternatives Considered

N/A

## Operations

- The authentication service storing users devices will need to be managed and probably be deployed separately from the Parsec metadata service
- GDPR will need to be considered for the data stored by the service. T&Cs will need to be adapted in order to mention the service and how to obtain and delete the data stored.

## Security/Privacy/Compliance

Consideration should be taken about the method used to save the device on the remote server.
The service would need to identify the user so we will need to have some information about the user (email mostly).

### Consideration on PBKDF algorithms

> 

Author: **todo:** Specific sub-derivation principle using `black2`
